sympy/solvers/tests/test_decompogen.py

```diff
@@ -19,7 +19,11 @@ def test_decompogen():
     assert decompogen(Abs(cos(y)**2 + 3*cos(x) - 4), x) == [Abs(x), 3*x + cos(y)**2 - 4, cos(x)]
     assert decompogen(x, y) == [x]
     assert decompogen(1, x) == [1]
+    assert decompogen(Max(3, x), x) == [Max(3, x), x]
     raises(TypeError, lambda: decompogen(x < 5, x))
+    u = 2*x + 3
+    raises(TypeError, lambda: decompogen(Max(u, u**2), x))
+    assert decompogen(Max(u, u**2, y), x) == [Max(x, y), 2*x + 3, x**2]
```

Comment:
this should be `[Max(x, x**2, y), 2*x + 3]` according to docstring:
```
Given an expression ``f``, returns a list ``[f_1, f_2, ..., f_n]``,
    where::
              f = f_1 o f_2 o ... f_n = f_1(f_2(... f_n))
    Note: This is a General decomposition function. It also decomposes
    Polynomials. For only Polynomial decomposition see ``decompose`` in polys.
    Examples
    ========
    >>> from sympy.abc import x
    >>> from sympy import decompogen, sqrt, sin, cos
    >>> decompogen(sin(cos(x)), x)
    [sin(x), cos(x)]
```
So if you keep replacing the left expression's x with the expression to the right you should get the original function.
`Min(x,x**2, y),x**2,2*x+3 -> Min(x, x**2, y) -> Min(2*x + 3, (2*x + 3)**2, y)`

Reply:
The logic should be something like
```python
args = list(f.args)
dlast = None
for i,a in enumerate(args):
    if not a.has_symbol: continue
    d = decompogen(a, symbol)
    if dlast is None:
        dlast = d
        continue
    if d[1:] != dlast[1:]:
        raise ValueError
    args[i] = d[0]
return [f.func(*args)] + d[1:]
```
So a test like `u = 2*x + 3; decompogen(Min(u, u**2, y, u**3), x)` should give `[Min(x, x**2, y, x**3), 2*x + y]` but there might be re-ordering of the args -- as long as the args are all there it should be fine.

(Really, it seems like `decompogen` should use `cse` so `decompogen(cos(2*x+3) + sin((2*x+3)**2 + y), x)` could give `cos(x) + sin(x**2 + y), 2*x + 3`. But that is a different issue.)

Reply:
`decompogen(Min(u, u**2, y, u**3), x)` is giving a ValueError for this change. Also `decompogen(Min(5, x**2), x)` now gives `[Min(5, x**2)]`.

Reply:
```
if all(a.has(symbol) for a in f.args):
            raise TypeError('cannot decompose %s' % f)
        args = list(f.args)
        d0 = None
        for i, a in enumerate(args):
            if not a.has(symbol):
                continue
            d = decompogen(a, symbol)
            if d0 is None:
                d0 = d
                continue
            args[i] = d[0]
        return [f.subs(args[1], symbol)] + d0
```
@smichr This change is working but I'm not sure whether it is the right way?